proposals/p2173.md

```diff
@@ -313,49 +317,165 @@ proposal suggests a simpler rule:
+interface A {
+  let T:! Type;
+  let U:! Type;
+}
+interface B {
+  extends A where .T = .U and .U = i32;
+}
+
+var n: i32;
+
+// ✅ Resolved constraint on `T` is
+// `B where .(A.T) = i32 and .(A.U) = i32`.
+// `T.(A.T)` is rewritten to `i32`.
+fn F(T:! B) -> T.(A.T) { return n; }
 ```
 
-In more detail, what happens here is:
+### Combining constraints with `impl as`
 
-1.  The constraint is fully elaborated. `.X` is replaced with `.Self.X`. For
-    example, `C where .T = .U and .U = i32` is rewritten to
-    `C where .T = .Self.U and .U = i32`.
-2.  When the constraint is applied to a type `V`, the substitution `.Self = V`
-    is performed. The type of type used for `V` is
-    `C where .T = <error> and .U = <error>`.
-3.  During this substitution, if either `<error>` rewrite rule would be used,
-    the result is an error. Substitution into `.Self.U` produces `V.U`, giving
-    an error.
+Within an interface or named constraint, the `impl T as C` syntax permits `=`
+constraints. However, because these constraints don't change the type of `T`,
+rewrite constraints will never result in a rewrite being performed, and instead
+are equivalent to `==` constraints.
+
+```
+interface A {
+  let T:! Type;
+  let U:! Type;
+}
+constraint C {
+  extends A where .T = .U and .U = i32;
+}
+constraint D {
+  extends A where .T == .U and .U == i32;
+}
+interface B {
+  // This constraint is equivalent to
+  // `impl as A where .T == .U and .U == i32;` or
+  // `impl as C;` or `impl as D;`.
+  impl as A where .T = .U and .U = i32;
+}
+
+var n: i32;
+
+// ❌ No implicit conversion from `i32` to `T.(A.T)`.
+// Resolved constraint on `T` is
+// `B where T.(A.T) == T.(A.U) and T.(A.U) = i32`.
+// `T.(A.T)` is single-step equal to `T.(A.U)`, and
+// `T.(A.U)` is single-step equal to `i32`, but
+// `T.(A.T)` is not single-step equal to `i32`.
+fn F(T:! B) -> T.(A.T) { return n; }
+```
+
+### Constraint resolution
+
+When a constraint is applied to a type, the constraint is _resolved_. This
+happens:
+
+-   When the constraint is used explicitly, when declaring a generic parameter
+    or associated constant of the form `T:! Constraint`.
+-   When declaring an `impl` such as `impl T as Constraint`.
+
+In each case, the following steps are performed to resolve the constraint value
+into a set of constraints on `T`:
+
+-   If multiple rewrites are specified for the same associated constant, they
+    are required to be identical, and duplicates are discarded.
+-   Rewrites are performed on other rewrites in order to find a fixed point,
+    where no rewrite applies within any other rewrite. If no fixed point exists,
+    the generic parameter declaration or `impl` declaration is invalid.
+-   Rewrites are performed throughout the other constraints in the constraint
+    value -- that is, in any `==` constraints and `is` constraints, and the type
+    `.Self` is replaced by `T` throughout the constraint.
+
+```
+// ✅ `.T` is rewritten to `i32` when initially
```

Comment:
Is `.U` also rewritten?
```suggestion
// ✅ `.T` and `.U` are rewritten to `i32` when initially
```

Reply:
No, this is specifically referring to the occurrence of `.T` in `.U = .T`.
```suggestion
// ✅ `.T` in `.U = .T` is rewritten to `i32` when initially
```

Reply:
(My suggestion has been applied; reopening so you can check if this resolves your comment.)